packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts

```diff
@@ -0,0 +1,494 @@
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+
+  /**
+   * Protocol the listener implements
+   */
+  protected protocol: Protocol = Protocol.HTTP;
+
+  constructor(props: HttpGatewayListenerProps = {}) {
+    super();
+    this.port = props.port ? props.port : 8080;
+    this.healthCheck = props.healthCheck;
+  }
+
+  /**
+   * Called when the GatewayListener type is initialized. Can be used to enforce
+   * mutual exclusivity
+   */
+  public bind(_scope: cdk.Construct): VirtualGatewayListenerConfig {
+    return {
+      listener: {
+        portMapping: {
+          port: this.port,
+          protocol: this.protocol,
+        },
+        healthCheck: renderHealthCheck(this.healthCheck, {
+          port: this.port,
+          protocol: this.protocol,
+        }),
+      },
+    };
+  }
+}
+
+/**
+* Represents the properties needed to define an HTTP2 Listener for a VirtualGateway
+*/
+class Http2GatewayListener extends HttpGatewayListener {
+  constructor(props: HttpGatewayListenerProps = {}) {
+    super(props);
+    this.protocol = Protocol.HTTP2;
+  }
+}
+
+/**
+ * Represents the properties needed to define a GRPC Listener for Virtual Gateway
+ */
+class GrpcGatewayListener extends VirtualGatewayListener {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port: number;
+
+  /**
+   * Health checking strategy upstream nodes should use when communicating with the listener
+   *
+   * @default - no healthcheck
+   */
+  readonly healthCheck?: HealthCheck;
+
+  /**
+   * Protocol the listener implements
+   */
+  protected protocol: Protocol = Protocol.GRPC;
+
+  constructor(props: HttpGatewayListenerProps = {}) {
+    super();
+    this.port = props.port ? props.port : 8080;
+    this.healthCheck = props.healthCheck;
+  }
+
+  /**
+   * Called when the GatewayListener type is initialized. Can be used to enforce
+   * mutual exclusivity
+   */
+  public bind(_scope: cdk.Construct): VirtualGatewayListenerConfig {
+    return {
+      listener: {
+        portMapping: {
+          port: this.port,
+          protocol: Protocol.GRPC,
+        },
+        healthCheck: renderHealthCheck(this.healthCheck, {
+          port: this.port,
+          protocol: Protocol.GRPC,
+        }),
+      },
+    };
+  }
+}
+
+/**
+ * Basic configuration properties for a VirtualGateway
+ */
+export interface VirtualGatewayBaseProps {
+  /**
+   * Name of the VirtualGateway
+   *
+   * @default - A name is automatically determined
+   */
+  readonly virtualGatewayName?: string;
+
+  /**
+   * Listeners for the VirtualGateway. Only one is supported.
+   *
+   * @default - Single HTTP listener on port 8080
+   */
+  readonly listeners?: VirtualGatewayListener[];
```

Comment:
Considering the condition checked below that a `VirtualGateway` can only have a single listener, why is this an array?

Reply:
We plan to support multiple listeners in the future

Reply:
> We're getting closer. I would still like to see a simplified version of the imported classes, and I'm very confused about `VirtualGateway` accepting multiple listeners, but then having validation that you can have only one...

We plan on supporting multiple listeners in the future: https://github.com/aws/aws-app-mesh-roadmap/issues/120

I went ahead and removed the `addListeners` methods since you cannot create a Virtual Gateway without a listener and since you cannot add more than one, those methods are useless.

Reply:
So you have to decide between one of two options here.

1. Because you only support a single listener right now, change this property to `readonly listener: VirtualGatewayListener`, and we'll add an additional property later that allows you to specify multiple ones.
2. Leave this as `readonly listeners: VirtualGatewayListener[]`, but remove the validation below that checks there's only one,  so that when you lift this limitation on the server side, CDK customers will be able to take advantage of it without the need to upgrade their CDK version.

Right now, this is a "worst of both worlds" kind of situation: you force providing an array, but don' allow specifying more than one because of the validation baked into the class. So the array serves no purpose from what I can see.

I'm leaving the decision to you which of the above options you want to go with 🙂.

Reply:
I'm leaning more towards option 2. 

Option 1 would require customers having to think about two separate fields, `listener` and `listeners` when creating a gateway/virtual node. 

Option 2 has the side affect of requiring server side validation, but you raise a good point on CDK versioning. I would prefer customers have immediate access when we change the server side validation.